pkg/bff/organization.go

```diff
@@ -13,6 +15,145 @@ import (
 	storeerrors "github.com/trisacrypto/directory/pkg/store/errors"
 )
 
+// CreateOrganization creates a new organization in the database. This endpoint returns
+// an error if the organization already exists and the user is assigned to it. The user
+// must have the create:organizations permission to perform this action.
+func (s *Server) CreateOrganization(c *gin.Context) {
+	var (
+		err  error
+		user *management.User
+	)
+
+	// Fetch the user from the context
+	if user, err = auth.GetUserInfo(c); err != nil {
+		log.Error().Err(err).Msg("create organization handler requires user info; expected middleware to return 401")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not identify user to create organization"))
+		return
+	}
+
+	// Load the user app metadata to check their organization assignments
+	appdata := &auth.AppMetadata{}
+	if err = appdata.Load(user.AppMetadata); err != nil {
+		log.Error().Err(err).Msg("could not parse user app metadata")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not parse user app metadata"))
+		return
+	}
+
+	// Unmarshal the params from the POST request
+	params := &api.OrganizationParams{}
+	if err := c.ShouldBind(params); err != nil {
+		log.Warn().Err(err).Msg("could not bind request")
+		c.JSON(http.StatusBadRequest, &api.Reply{Error: err.Error()})
+		return
+	}
+
+	// Name is a required paramater
+	if params.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("must provide name in request params"))
+		return
+	}
+
+	// Domain is a required paramater
+	domain := normalizeDomain(params.Domain)
+	if domain == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("must provide domain in request params"))
+		return
+	}
+
+	// Don't allow the user to create duplicate organizations
+	// TODO: Should we do a universal check against the database using an index?
+	for _, id := range appdata.GetOrganizations() {
+		if org, err := s.OrganizationFromID(id); err != nil {
+			log.Error().Err(err).Str("org_id", id).Msg("could not retrieve organization from database")
+		} else if org.Domain == domain {
+			c.JSON(http.StatusConflict, api.ErrorResponse("organization with domain name already exists"))
+			return
+		}
+	}
+
+	// Create a new organization in the database
+	var org *models.Organization
+	if org, err = s.db.CreateOrganization(); err != nil {
+		log.Error().Err(err).Msg("could not create organization in database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create organization"))
+		return
+	}
+
+	// Update the organization with the request params
+	org.Name = params.Name
+	org.Domain = domain
+	if err = s.db.UpdateOrganization(org); err != nil {
+		log.Error().Err(err).Msg("could not update organization in database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create organization"))
+		return
+	}
```

Comment:
Are two DB queries necessary? What is `CreateOrganization` doing that requires us to update the org and then save it again? E.g. can we get name and domain into the organization at create time?

Reply:
I think it's definitely possible to do that, we just have to add or edit a trtl method. This is the first time we've known anything about the organization before creating it since we were previously just creating them for new users.

Reply:
Ok, I think that might be useful to do if that's ok with you.